cradle-core/src/main/java/com/exactpro/cradle/messages/StoredGroupedMessageBatch.java

```diff
@@ -129,4 +127,32 @@ protected TreeSet<StoredMessage> createMessagesCollection(Collection<StoredMessa
 			result.add(new StoredMessage(msg, msg.getId(), pageId));
 		return result;
 	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) return true;
+		if (!(o instanceof StoredGroupedMessageBatch)) return false;
+		StoredGroupedMessageBatch that = (StoredGroupedMessageBatch) o;
+		return getBatchSize() == that.getBatchSize()
+				&& getBookId().equals(that.getBookId())
+				&& getGroup().equals(that.getGroup())
+				&& getMessages().equals(that.getMessages())
+				&& getRecDate().equals(that.getRecDate());
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(getBookId(), getGroup(), getBatchSize(), getMessages(), getRecDate());
```

Comment:
How many times we use one object as key? 
We can write hash into class variable if we calculate hash for each instances 1 or more times